### PR TITLE
ISPN-14872 Include RESP in AbstractAuthorization tests

### PR DESCRIPTION
**Pull request by jabolina**
https://issues.redhat.com/browse/ISPN-14872
https://issues.redhat.com/browse/ISPN-14895

Besides adding the tests with the server authorization, this also includes a way to authenticate using a certificate. The only thing not covered is Kerberos, which is skipped.

### CODE REVIEW

File: server/resp/src/main/java/org/infinispan/server/resp/Resp3AuthHandler.java [hunk 1]
```diff
@@ -38,12 +39,29 @@ public CompletionStage<Boolean> performAuth(ChannelHandlerContext ctx, byte[] us
...
+      return performAuth(ctx, (String) null, null);
+   }
+
    private CompletionStage<Boolean> performAuth(ChannelHandlerContext ctx, String username, String password) {
-      UsernamePasswordAuthenticator authenticator = respServer.getConfiguration().authentication().authenticator();
+      RespAuthenticator authenticator = respServer.getConfiguration().authentication().authenticator();
       if (authenticator == null) {
-         return CompletableFutures.booleanStage(handleAuthResponse(ctx, null));
+         boolean success = handleAuthResponse(ctx, null);
+         return success ? CompletableFutures.completedTrue() : CompletableFutures.completedFalse();
```
Comment (tristantarrant) [inline server/resp/src/main/java/org/infinispan/server/resp/Resp3AuthHandler.java:50]: The original code was better :)

File: server/resp/src/main/java/org/infinispan/server/resp/Resp3AuthHandler.java [hunk 2]
```diff
@@ -38,12 +39,29 @@ public CompletionStage<Boolean> performAuth(ChannelHandlerContext ctx, byte[] us
...
+      }
+
+      CompletionStage<Subject> authentication;
+      if (username == null && password == null) {
```
Comment (tristantarrant) [inline server/resp/src/main/java/org/infinispan/server/resp/Resp3AuthHandler.java:54]: Probably you should also check `authenticator.isClientCertAuthEnabled()` here

File: server/runtime/src/main/java/org/infinispan/server/configuration/endpoint/EndpointConfigurationBuilder.java [hunk 3]
```diff
@@ -272,17 +274,31 @@ public static void enableImplicitAuthentication(SecurityConfiguration security,
...
+         respAuthenticator.withClientCertAuth(new RespClientCertAuthenticator());
+         authSupported = true;
+      }
+
+      if (!authSupported) {
          if (builder.implicitConnector()) {
             // The connector was added implicitly, but the security realm cannot support it. Remove it.
             return null;
          } else {
             throw Server.log.respEndpointRequiresRealmWithPassword();
```
Comment (tristantarrant) [inline server/runtime/src/main/java/org/infinispan/server/configuration/endpoint/EndpointConfigurationBuilder.java:297]: Change the log to also say "or a truststore realm"

File: server/runtime/src/main/java/org/infinispan/server/security/RespClientCertAuthenticator.java [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+import org.infinispan.server.core.security.simple.SimpleSaslAuthenticator;
+
+import io.netty.channel.Channel;
+import io.netty.handler.ssl.SslHandler;
+
+public class RespClientCertAuthenticator {
+
+
+   public CompletionStage<Subject> clientCertAuth(Channel channel) throws SaslException {
+      SaslAuthenticator sap = new SimpleSaslAuthenticator();
```
Comment (tristantarrant) [inline server/runtime/src/main/java/org/infinispan/server/security/RespClientCertAuthenticator.java:27]: You should not be using the `SimpleSaslAuthenticator` but rather rely on the Elytron SASL authenticator which has been configured with a `TrustStoreRealm` and therefore can validate the client certificates against the trust store.

**Review by tristantarrant — CHANGES_REQUESTED** [review]
I would really like to split `AbstractAuthorization` into per-protocol classes and turn the various `*AuthorizationIT` classes into `Suite`s.

**Comment by jabolina** [comment]
Applied all the suggestions. Since we have the command `INFO`, I added a test for it, meaning slight changes, storing the subject with an attribute key and the permission check.

Now, I'll focus on trying to split the tests per protocol. But the other parts are ready for review again.

File: server/resp/src/main/java/org/infinispan/server/resp/Resp3Handler.java [hunk 5]
```diff
@@ -76,10 +81,7 @@ public CompletionStage<RespRequestHandler> delegate(ChannelHandlerContext ctx,
       return super.actualHandleRequest(ctx, command, arguments);
    }
 
-   public void checkPermission(AuthorizationPermission authorizationPermission) {
-      AuthorizationManager authorizationManager = cache.getAuthorizationManager();
-      if (authorizationManager != null) {
-         authorizationManager.checkPermission(authorizationPermission);
-      }
+   public void checkPermission(Subject subject, AuthorizationPermission authorizationPermission) {
+      SecurityActions.checkPermission(respServer.getCacheManager(), subject, authorizationPermission);
```
Comment (tristantarrant) [inline server/resp/src/main/java/org/infinispan/server/resp/Resp3Handler.java:85]: This code is wrong. 
Some authz checks need to be done against the cache permissions, while others may need to be done using the global permissions.

**Comment by jabolina** [comment]
Fixed the previous issues. Some resources were leaking after the test was finished. But now Lettuce threads are leaking.

File: server/resp/src/main/java/org/infinispan/server/resp/Resp3AuthHandler.java [hunk 6]
```diff
@@ -5,14 +5,17 @@
...
 import org.infinispan.commons.util.concurrent.CompletableFutures;
-import org.infinispan.server.core.security.UsernamePasswordAuthenticator;
+import org.infinispan.server.resp.authentication.RespAuthenticator;
 import org.infinispan.server.resp.commands.AuthResp3Command;
 
 import io.netty.channel.ChannelHandlerContext;
+import io.netty.util.AttributeKey;
 
 public class Resp3AuthHandler extends CacheRespRequestHandler {
+   static final AttributeKey<Subject> SUBJECT_ATTRIBUTE_KEY = AttributeKey.newInstance("subject");
```
Comment (tristantarrant) [inline server/resp/src/main/java/org/infinispan/server/resp/Resp3AuthHandler.java:18]: @jabolina I'm adding the Subject to the `ConnectionMetadata` in https://github.com/infinispan/infinispan/pull/10964

**Comment by jabolina** [comment]
Updated removing the Subject attr key. The last CI run had a [failure in the REST backup test](https://ci.infinispan.org/job/Infinispan/job/PR-10961/8/testReport/org.infinispan.server.security.authorization/RESTAuthorizationTest/testRestAdminsMustAccessBackupsAndRestores/).

This was triggered by running the RESP tests first, and since the `respCache` would have heterogeneous entries, plain `byte[]` and multimap objects, the test would fail. I added a little workaround checking the metadata, but that's until we establish a way to verify the entries properly for data structures.

**Comment by jabolina** [comment]
Looks like I broke everything again.

**Comment by jabolina** [comment]
Well, the thread dump had almost 300 Ok Http threads, and the CI logs have some hints about that, too. I took some time removing all usages of `sync` in the REST tests, which could leak some resources. Let's see how it goes.

**Comment by jabolina** [comment]
Looks like the failures are unrelated.

**Review by tristantarrant — APPROVED** [review]

**Comment by tristantarrant** [comment]
Merged, thanks

**Pull request merged**